saq/queue/postgres.py

```diff
@@ -0,0 +1,608 @@
+                # Could not acquire the sweep lock so another worker must already be sweeping
+                return []
+            await cursor.execute(
+                SQL(
+                    """
+                DELETE FROM {jobs_table}
+                WHERE status IN ('aborted', 'complete', 'failed')
+                  AND %(now)s >= ttl
+                RETURNING key
+                """
+                ).format(jobs_table=self.jobs_table),
+                {"now": math.ceil(seconds(now()))},
+            )
+            await cursor.execute(
+                SQL(
+                    """
+                SELECT job
+                FROM {jobs_table} LEFT OUTER JOIN pg_locks ON lock_key = objid
+                WHERE status = 'active'
+                  AND objid IS NULL
+                FOR UPDATE SKIP LOCKED
+                """
+                ).format(jobs_table=self.jobs_table)
+            )
+            results = await cursor.fetchall()
+            for result in results:
+                job = self.deserialize(result[0])
+                if not job:
+                    continue
+                swept.append(job.key)
+                if job.retryable:
+                    await self.retry(job, error="swept")
+                else:
+                    await job.finish(Status.ABORTED, error="swept")
+        return swept
+
+    async def listen(
+        self,
+        job_keys: Iterable[str],
+        callback: ListenCallback,
+        timeout: float | None = 10,
+    ) -> None:
+        if not job_keys:
+            return
+
+        async with self.pool.connection() as conn:
+            for key in job_keys:
+                await conn.execute(f'LISTEN "{key}"')
+            await conn.commit()
+            gen = conn.notifies(timeout=timeout or None)
+            async for notify in gen:
+                payload = self._load(notify.payload)
+                key = payload["key"]
+                status = Status[payload["status"].upper()]
+                if asyncio.iscoroutinefunction(callback):
+                    stop = await callback(key, status)
+                else:
+                    stop = callback(key, status)
+
+                if stop:
+                    await gen.aclose()
+
+    async def notify(self, job: Job) -> None:
+        payload = self._dump({"key": job.key, "status": job.status})
+        await self._notify(job.key, payload)
+
+    async def _notify(self, channel: str, payload: t.Any) -> None:
+        async with self.pool.connection() as conn:
+            await conn.execute(f"NOTIFY \"{channel}\", '{payload}'")
+
+    async def update(
+        self, job: Job, status: Status | None = None, scheduled: int | None = None
+    ) -> None:
+        if status:
+            job.status = status
+        if scheduled:
+            job.scheduled = scheduled
+        job.touched = now()
+
+        async with self.pool.connection() as conn:
+            await conn.execute(
```

Comment:
is this done in a transaction? can we also notify in the transaction too so that you'll never update a job and skip a notification?

Reply:
can you look through all the places and make sure things are as robust as possible with using transactions?